### PR TITLE
Integrate mHC with DeepSeek custom model

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

* Update default mhc expansion rate to 1, which is the same as disable the feature.
* Update the shape of activations in `decoders.py` when feature is enabled.
* Enable the loss tracking in MoE layers when using mHC.
* Update the precision of weights to activation before matmul, which aligns with existing pattern in MaxText

General pre-norm when mHC feature is disabled:

```
Input (x) ───────────────────────────┐
  │                                  │
  ▼                                  │ (Residual Connection)
[ Pre-Norm ]                         │
  │                                  │
  ▼                                  │
[ Attention / MLP ]                  │
  │                                  │
  ▼                                  │
[ Layer Output ]                     │
  │                                  │
  └───────────► ( + ) ◄──────────────┘
                 │
                 ▼
           Final Output

```

When mHC feature is enabled: 

```

Input (x) ─────────────────────────-─────────┐
          │                                  │
          │    Pre mapping                   |  
          │    pre-norm                      |  
          ▼                           residual mapping     
  [ Attention / MLP ]                        │      
          │                                  |
          ▼                                  │   
  [ Layer Output ]                           │          
          │   post mapping                   │             
          |                                  │         
         layer_output                      res_output           
          └──────────► ( + ) ◄────────--─────┘          
                        │                             
                        ▼                             
                  Final Output                 

```

# Tests

* Update unit tests
* End-to-end sanity check test - [link](https://paste.googleplex.com/4918137587892224)
* Check MoE related load balance is captured in TB - [link](https://screenshot.googleplex.com/7tNYs9rHBZWpGcF)
* Check mHC end-to-end (500 steps with seed datasets) - [comparison](https://tensorboard.corp.google.com/compare/with-mhc:8802821794889206282,no-mhc:1847534110015664336/?darkMode=true#timeseries). Please note, the paper is comparing mHC vs. HC. Here is comparing mHC vs. baseline.
   * mHC - loss: 6.701 (expansion_rate=4), slightly lower on this toy model with real dataset
   * Without mHC - loss: 6.796 (expansion_rate=1)

```
# cmd to run

python3 -m MaxText.train maxtext/configs/base.yml base_output_directory=${BASE_OUTPUT_PATH} run_name=${RUN_NAME} per_device_batch_size=8 enable_checkpointing=false model_name=deepseek-custom ici_fsdp_parallelism=4 steps=500 max_target_length=4096 async_checkpointing=false dtype=bfloat16 weight_dtype=float32 scan_layers=True dataset_type=synthetic attention=dot_product train_split=train dataset_type=hf hf_path='HuggingFaceFW/fineweb-edu' hf_name=default enable_tensorboard=true tokenizer_type=huggingface tokenizer_path=deepseek-ai/DeepSeek-V3.2 data_shuffle_seed=1234
```

* DeepSeek v2 sanity tests (expect no impact for existing models)

```
python3 -m MaxText.train maxtext/configs/base.yml base_output_directory=${BASE_OUTPUT_PATH} run_name=${RUN_NAME} per_device_batch_size=8 enable_checkpointing=false model_name=deepseek2-16b ici_fsdp_parallelism=4 steps=20 max_target_length=4096 async_checkpointing=false tokenizer_path=src/MaxText/assets/tokenizer.mistral-v1 dtype=bfloat16 weight_dtype=float32 scan_layers=True dataset_type=synthetic attention=flash

# before change

I0210 00:18:46.050221 139944566173248 metric_logger.py:181] completed step: 19, seconds: 4.363, TFLOP/s/device: 123.215, Tokens/s/device: 7510.083, total_weights: 131072, loss: 8.135

# after change

I0210 00:09:42.323213 139931725155904 metric_logger.py:181] completed step: 19, seconds: 4.365, TFLOP/s/device: 123.166, Tokens/s/device: 7507.116, total_weights: 131072, loss: 8.135

```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/layers/deepseek.py [hunk 1]
```diff
@@ -172,31 +179,21 @@ def attention_op(
 
   @property
   def logical_axis_names(self):
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/MaxText/layers/deepseek.py:179]: Will update to None --> "mhc" when Engram PR is merged.

File: src/MaxText/layers/decoders.py [hunk 2]
```diff
@@ -731,6 +731,11 @@ def __call__(
         audio_masks,
     )
 
+    if cfg.mhc_expansion_rate > 1:
+      # (batch, length, emb_dim) --> (batch, length, mhc_expansion_rate, emb_dim)
+      y = y[:, :, None, :]
+      y = jnp.repeat(y, cfg.mhc_expansion_rate, axis=2).astype(y.dtype)
```
Comment (shuningjin) [inline src/MaxText/layers/decoders.py]: reuse expand and reduce?
https://github.com/AI-Hypercomputer/maxtext/blob/50ef2df481aa1084f031e0f6fb9b9122b419de0c/src/MaxText/layers/mhc.py#L35-L41
https://github.com/AI-Hypercomputer/maxtext/blob/b6e0cdb6b204ffa2f782e8c93c4ab527f4e3c114/tests/unit/mhc_test.py#L36

File: src/MaxText/layers/mhc.py [hunk 3]
```diff
@@ -162,74 +164,88 @@ def __init__(
...
-        sharding=(None, None),
+        sharding=(None,),
     )
 
   def res_mapping(self, x: Array):
     """Helper function for residual mapping."""
+    # In MaxText, we match weight precision to activations before Matmul
+    res_alpha = jnp.asarray(self.res_alpha[...], self.dtype)
+    res_beta = jnp.asarray(self.res_beta[...], self.dtype)
+    res_alpha_scale = jnp.asarray(self.res_alpha_scale, self.dtype)
```
Comment (shuningjin) [inline src/MaxText/layers/mhc.py]: nit: maybe use [...] for scale as well for consistency? I see moe bias https://github.com/AI-Hypercomputer/maxtext/blob/50ef2df481aa1084f031e0f6fb9b9122b419de0c/src/MaxText/layers/moe.py#L296
potentially a function for these
```
def cast(weight):
  return jnp.asarray(weight[...], self.dtype)
```

File: src/MaxText/layers/deepseek.py [hunk 4]
```diff
@@ -43,6 +44,10 @@
 # -----------------------------------------
 
 
+def is_mhc_enabled(expansion_rate):
```
Comment (shuningjin) [inline src/MaxText/layers/deepseek.py]: nit: make this an attribute (self.is_mhc_enable)?

File: src/MaxText/layers/mhc.py [hunk 5]
```diff
@@ -162,74 +164,88 @@ def __init__(
...
       mhc_type: HyperConnectionType,
       **kwargs,
   ) -> Array:
     """Applying manifold-constrained hyper connection based on callable function.
 
     Args:
         branch_fn: The function to be wrapped by the hyper-connection.
-        x: Input tensor of shape `(batch..., dim)`.
+        residual_x: Input tensor of shape `(batch, seq, expansion_rate, dim)`, generally before pre-normalization.
+        layer_x: Input tensor of shape `(batch, seq, expansion_rate, dim)`, generally after pre-normalization.
```
Comment (shuningjin) [inline src/MaxText/layers/mhc.py]: I didn't understand this part. Is it using pre-norm on top of mhc? Could you explain more?

**Review by shuningjin — COMMENTED** [review]
Thanks for the integration and testing! 

> Update mHC a little bit to align with normalization (flexible with pre-norm and post-norm). Please note mHC also has a norm across last k * dim dimension inside.

I’m not sure I fully follow this part. My understanding was that HC is designed to replace pre-norm and post-norm: HC subsumes both forms and overcomes their limitation, [ref](https://screenshot.googleplex.com/4UmASywPvZTfesD). Great visualization, I think "Pre mapping" should replace "pre-norm", rather than together. Appreciate your clarification!

File: src/MaxText/layers/mhc.py (hunk 5 above)
Comment (RissyRan) [reply src/MaxText/layers/mhc.py]: Let's discuss offline.

File: src/MaxText/layers/deepseek.py (hunk 4 above)
Comment (RissyRan) [reply src/MaxText/layers/deepseek.py]: Updated and tested.

File: src/MaxText/layers/decoders.py (hunk 2 above)
Comment (RissyRan) [reply src/MaxText/layers/decoders.py]: Good call. Updated and tested.

File: src/MaxText/layers/mhc.py (hunk 3 above)
Comment (RissyRan) [reply src/MaxText/layers/mhc.py]: Updated and tested.

**Comment by RissyRan** [comment]
> Thanks for the integration and testing!
> 
> > Update mHC a little bit to align with normalization (flexible with pre-norm and post-norm). Please note mHC also has a norm across last k * dim dimension inside.
> 
> I’m not sure I fully follow this part. My understanding was that HC is designed to replace residual connection (either pre-norm and post-norm form), rather than co-exist. HC subsumes both forms and overcomes their limitation, [ref](https://screenshot.googleplex.com/4UmASywPvZTfesD).
> 
> [Update] I need another look at mhc paper, [this](https://screenshot.googleplex.com/AEaeKLgazNoh43y)

I think that's right. HC has learnable weights and scale to make flexible as pre/post-norm. This makes it much easier! Let me update it back to original implementation.

**Comment by RissyRan** [comment]
> > Thanks for the integration and testing!
> > > Update mHC a little bit to align with normalization (flexible with pre-norm and post-norm). Please note mHC also has a norm across last k * dim dimension inside.
> > 
> > 
> > I’m not sure I fully follow this part. My understanding was that HC is designed to replace residual connection (either pre-norm and post-norm form), rather than co-exist. HC subsumes both forms and overcomes their limitation, [ref](https://screenshot.googleplex.com/4UmASywPvZTfesD).
> > [Update] I need another look at mhc paper, [this](https://screenshot.googleplex.com/AEaeKLgazNoh43y)
> 
> I think that's right. HC has learnable weights and scale to make flexible as pre/post-norm. This makes it much easier! Let me update it back to original implementation.

Discussed offline. There are misalignment between DS mHC and original HC implementation (HC has [pre-norm](https://screenshot.googleplex.com/3Q7f2gEniQ6WuTF), weights shape are different to mHC, and HC does not utilize mapping for residuals) . We would like to wait for the clarification when official implementation is available. Also added a comment in the codebase.

```
  Note: As an official reference implementation is currently unavailable, the
  integration of pre-norm and post-norm within the mHC architecture remains
  to further verification.
```

File: src/MaxText/layers/decoders.py [hunk 6]
```diff
@@ -731,6 +732,11 @@ def __call__(
         audio_masks,
     )
 
+    mhc_expand, mhc_reuce = mhc.get_functions(cfg.mhc_expansion_rate)
```
Comment (shuningjin) [inline src/MaxText/layers/decoders.py]: nit: mhc_reuce -> mhc_reduce

File: src/MaxText/layers/deepseek.py [hunk 7]
```diff
@@ -233,16 +228,30 @@ def self_attention_with_norm_op(
...
-        slot,
-    )
-    intermediate_inputs = inputs + attention_lnx
+    if self.is_mhc_enabled:
+      intermediate_inputs, _ = self.mhc_attention(
+          self.self_attention,
+          x=inputs,
+          mhc_type=HyperConnectionType.ATTENTION,
+          decoder_segment_ids=decoder_segment_ids,
+          deterministic=deterministic,
```
Comment (gagika) [inline src/MaxText/layers/deepseek.py:233]: should you also pass decoder_positions, as MLA depends on it?

inputs_positions=decoder_positions,

File: src/MaxText/layers/decoders.py [hunk 8]
```diff
@@ -731,6 +732,11 @@ def __call__(
         audio_masks,
     )
 
+    mhc_expand, mhc_reduce = mhc.get_functions(cfg.mhc_expansion_rate)
```
Comment (gagika) [inline src/MaxText/layers/decoders.py:735]: nit: if mhc_reduce not used, use  mhc_expand, _ = mhc.get_functions

File: src/MaxText/layers/deepseek.py [hunk 9]
```diff
@@ -233,16 +228,30 @@ def self_attention_with_norm_op(
     """self-attention with normalization"""
     lnx = self.pre_attention_norm_op(inputs)
```
Comment (gagika) [inline src/MaxText/layers/deepseek.py]: if it's not used in if self.is_mhc_enabled brunch we can move it in else.

Based on the paper Gemini thinks right way to do it is:


```
def self_attention_with_norm_op(
      self,
      inputs,
      decoder_segment_ids,
      decoder_positions,
      deterministic,
      previous_chunk=None,
      page_state: None | page_manager.PageState = None,
      slot: None | int = None,
  ):
    """self-attention with normalization"""

    if self.is_mhc_enabled:
      # Create a wrapper function that perfectly mimics the paper's math: RMSNorm(H^pre * x_l)
      def wrapped_attention_fn(inputs_q, inputs_kv, **kwargs):
          # 1. Apply the Pre-Attention Norm to the mapped inputs (inputs_q)
          lnx = self.pre_attention_norm_op(inputs_q)
          
          # 2. Pass the normalized inputs into the standard attention_op
          return self.attention_op(
              lnx,
              decoder_segment_ids=kwargs.get("decoder_segment_ids"),
              decoder_positions=kwargs.get("inputs_positions"),
              deterministic=kwargs.get("deterministic"),
              previous_chunk=kwargs.get("previous_chunk"),
              page_state=kwargs.get("page_state"),
              slot=kwargs.get("slot"),
          )

      # Pass the UN-NORMALIZED inputs to mHC, just as the paper specifies
      intermediate_inputs = self.mhc_attention(
          wrapped_attention_fn,
          x=inputs, 
          mhc_type=HyperConnectionType.ATTENTION,
          decoder_segment_ids=decoder_segment_ids,
          inputs_positions=decoder_positions,
          deterministic=deterministic,
          previous_chunk=previous_chunk,
          page_state=page_state,
          slot=slot,
      )
    else:
      lnx = self.pre_attention_norm_op(inputs)
      attention_lnx = self.attention_op(
          lnx,
          decoder_segment_ids,
          decoder_positions,
          deterministic,
          previous_chunk,
          page_state,
          slot,
      )
      intermediate_inputs = inputs + attention_lnx
      
    # Normalization
    hidden_states = self.post_attention_norm_op(intermediate_inputs)
    return hidden_states, intermediate_inputs
```

e.g. the input of `mhc_attention` is non-normalized inputs, but `wrapped_attention_fn` does normalization before calling `self.attention_op`

     intermediate_inputs = self.mhc_attention(
          wrapped_attention_fn,
          x=inputs,

File: src/MaxText/layers/decoders.py (hunk 8 above)
Comment (RissyRan) [reply src/MaxText/layers/decoders.py:735]: mhc_reduce is used later :) https://screenshot.googleplex.com/C42RoiJKconDMQG

File: src/MaxText/layers/deepseek.py (hunk 7 above)
Comment (RissyRan) [reply src/MaxText/layers/deepseek.py:233]: Good catch! Updated.

File: src/MaxText/layers/deepseek.py (hunk 9 above)
Comment (RissyRan) [reply src/MaxText/layers/deepseek.py]: Thanks! Discussed offline, and updated to current implementation to have pre-norm inside of mHC.

**Comment by shuningjin** [comment]
Thanks for the update, LGTM!

I have high confidence that current implementation aligns with both papers.
- checked MHC paper's equation about MHC ([paper](https://arxiv.org/pdf/2512.24880))
- checked HC paper's pseudo code of HC ([HyerConnection paper Appendix J](https://arxiv.org/pdf/2409.19606)). In particular, this reference informs the use of pre-norm.
- summarized my finding in https://b.corp.google.com/issues/478294028#comment6.

**Review by gagika — APPROVED** [review]
thanks
